File: 1798-max-number-of-k-sum-pairs/1798-max-number-of-k-sum-pairs.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxOperations = function(nums, k) {
    const map = new Map();
    let count = 0;

    for (let i = 0; i < nums.length; i++) {
        if (map.has(k - nums[i]) && map.get(k - nums[i]) > 0) {
            map.set(k - nums[i], map.get(k - nums[i]) - 1);
            count++;
        } else {
            map.set(nums[i], (map.get(nums[i]) || 0) + 1);
        }
    }

    return count;

};

File: 1798-max-number-of-k-sum-pairs/1798-max-number-of-k-sum-pairs.cpp
class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        unordered_map<int , int> map;

        int count =0;

        for(int i=0; i<nums.size(); i++){
            if(map[k-nums[i]] && map[k-nums[i]] > 0){
                map[k-nums[i]]--;
                count++;
            }else{
                map[nums[i]]++;
            }
        }
        return count;
    }
};